Fix coef and sign

razuf opened the pull request:
Hi,

I found a bug or structural problem in Decimal what could have some big impact under special circumstances. Here an example from the Money lib where Decimal is used as main backbone:

```elixir
iex(15)> m = Money.new(:USD, Decimal.new(%Decimal{sign: -1, coef: -12000000000}))
#Money<:USD, --12000000000>
iex(16)> Money.mult(m, 3)
{:ok, #Money<:USD, --36000000000>} 
iex(17)> Money.mult!(m, 3 )|> Money.to_string 
{:ok, "$--*,000,000,000.00"}
```

Of course it's a special edge case ... normally the best practice is:

```elixir
iex(18)> m = Money.new(:USD, "-12000000000") 
#Money<:USD, -12000000000>
```
but you never know ... maybe someone using it ...it's possible! it's valid Code ! 
NO compiler warning! (also true if you are using dialyzer, when using var names for input)
NO runtime error!

And the worst case, if you are using any division with this kind of Decimal number like this:

```elixir
iex(19)> m = Money.new(:USD, Decimal.new(%Decimal{sign: -1, coef: -12000000000}))
#Money<:USD, --12000000000>
iex(20)> Money.mult!(m, 3) |> Money.to_decimal |> Decimal.to_float
```
-> you end up in an infinite loop!!! 

The reason for this is clear: the %Decimal{} struct - with not enough validation for input -> so you can input a negative coefficient.

The type spec and docs are perfect - but do not prevent any misuse: 
Type spec :  
 @type coefficient :: non_neg_integer | :NaN | :inf
  @type sign :: 1 | -1

Documentation: The coefficient of the power of `10`. Non-negative because the sign is stored separately in `sign`.

So my suggestion for a solution you can find in this pull request: mostly checks for this special edge case.

And this time I included the corresponding tests. ;-) 

one example for the additional function clause (now backwards compatible to Elixir 1.2):
```elixir
def add(%Decimal{coef: coef1}, %Decimal{coef: coef2})
      when coef1 < 0 or coef2 < 0 do
    error(:invalid_operation, "wrong coef", %Decimal{coef: :NaN})
  end

  def add(%Decimal{sign: sign1}, %Decimal{})
      when sign1 != 1 and sign1 != -1 do
    error(:invalid_operation, "wrong sign", %Decimal{coef: :NaN})
  end

  def add(%Decimal{}, %Decimal{sign: sign2})
      when sign2 != 1 and sign2 != -1 do
    error(:invalid_operation, "wrong sign", %Decimal{coef: :NaN})
  end
```

The test and benchees are fine - no big performance cost - the differences are marginal (see deviation):

before the fix:

```elixir
##### With input 30 Digits Integer #####
Name                           ips        average  deviation         median         99th %
Decimal.add()    Int      318.01 K        3.14 μs   ±771.75%           3 μs           5 μs

Memory usage statistics:

Name                    Memory usage
Decimal.add()    Int         4.26 KB

**All measurements for memory usage were the same**

```

with the fix:
```elixir
##### With input 30 Digits Integer #####
Name                           ips        average  deviation         median         99th %
Decimal.add()    Int      316.97 K        3.15 μs   ±659.73%           3 μs           5 μs

Memory usage statistics:

Name                    Memory usage
Decimal.add()    Int         4.26 KB

**All measurements for memory usage were the same**

```

razuf commented:
Oh yeah - sorry it failed on old elixir versions - I will fix it.

ericmj commented:
It would be hard to catch every scenario where a user constructs an invalid `Decimal` struct. They could for example pass a float in any of the fields and the library would also work incorrectly. Adding guards to every function that catches every invalid value seems fragile and not very feasible. In addition we are catching the error too late, at the time where the struct is passed to the function instead of when the invalid struct is created.

It is better to use the provided functions such as `Decimal.new/1` to create decimals than to create them manually. In the past the `Decimal.t/0` type was opaque to prevent users from creating structs manually, but libraries such as Ecto/Postgrex needed to create structs manually when casting the values from postgres' wire format. I think that's one of the only times where you should not (cannot) create the structs manually.

I appreciate the PR, but because of these reasons I am hesitant to merge it.

wojtekmach commented:
We already have an `is_decimal/1` macro, I believe it would be possible to extend to check the struct more closely (e.g. `%Decimal{coef: coef} when is_integer(coef) and coef >= 0` and the equivalent for guards) and use it for existing functions.

That being said, as mentioned in this issue and [elixirforum](https://elixirforum.com/t/decimal-bug-please-be-careful-with-negative-coefficients-fix-is-on-pull-request/32430), it is more common to assume that a given struct was properly constructed in the first place and skip further checks (that's why they are _structs_ and not merely _maps_). I believe extending the is_decimal/1 macro further would have small additional runtime cost too but perhaps it's negligible anyway.

razuf commented:
Many thanks for reviewing PR and for your comment. I appreciate that.

I can follow that argument:

> Adding guards to every function that catches every invalid value seems fragile and not very feasible.

and too: 
> we are catching the error too late

thats why I had the check in new/1 and cast/1 - but what do you think about the adapted PR with only checking that (I removed the rest): 

For the inbound:
new/1 - is the same check as in new/3 
```elixir
  def new(%Decimal{sign: sign, coef: coef, exp: exp} = num)
      when sign in [1, -1] and ((is_integer(coef) and coef >= 0) or coef in [:NaN, :inf]) and
             is_integer(exp) do
    num
  end
```

and added a default case error (in my opinion better error handling then FunctionClauseError - but I can live without that..)

```elixir
  def new(num) do
    raise Error, reason: "wrong decimal number: #{inspect(num)}"
  end

```

cast - I think that's the behaviour from cast I would expect - almost always :error instead of "let it crash":

For the outbound:
to_string, to_integer, to_float

That's fixing most of the tests in the decimal_test.ex from before - and increase the security of the Decimal lib - I think" ;-)

one of the edge case did not fixed would be only:

```elixir
test "test sqrt with wrong sign -> infinite loop" do
  assert_raise Decimal.Error, fn ->
    Decimal.sqrt(d(3, 1, -1))
  end
end

```

As you said - we cannot handle all possible cases...

But the Money examples from above would be fixed by this PR.
